examples/platforms/efr32/radio.c

```diff
@@ -760,14 +792,44 @@ static void RAILCb_Generic(RAIL_Handle_t aRailHandle, RAIL_Events_t aEvents)
         status = RAIL_Calibrate(aRailHandle, NULL, RAIL_CAL_ALL_PENDING);
         assert(status == RAIL_STATUS_NO_ERROR);
     }
+
+    if (aEvents & RAIL_EVENT_RSSI_AVERAGE_DONE)
+    {
+        const int16_t energyScanResultQuarterDbm = RAIL_GetAverageRssi(aRailHandle);
+
+        if (energyScanResultQuarterDbm == RAIL_RSSI_INVALID)
+        {
+            sEnergyScanResultDbm = 0;
+            sEnergyScanStatus    = ENERGY_SCAN_COMPLETED_ERROR;
+        }
+        else
+        {
+            sEnergyScanResultDbm = energyScanResultQuarterDbm / QUARTER_DBM_IN_DBM;
+            sEnergyScanStatus    = ENERGY_SCAN_COMPLETED;
+        }
+    }
 }
 
 otError otPlatRadioEnergyScan(otInstance *aInstance, uint8_t aScanChannel, uint16_t aScanDuration)
 {
+    RAIL_Status_t status;
+    otError       error = OT_ERROR_NONE;
+
     (void)aInstance;
-    (void)aScanChannel;
-    (void)aScanDuration;
-    return OT_ERROR_NOT_IMPLEMENTED;
+
+    CORE_DECLARE_IRQ_STATE;
+    CORE_ENTER_CRITICAL();
+
+    otEXPECT_ACTION((sEnergyScanStatus == ENERGY_SCAN_IDLE && RAIL_GetRadioState(sRailHandle) == RAIL_RF_STATE_IDLE),
```

Comment:
Does this allow OpenThread to perform energy scan even while operating in the router/leader role?

Reply:
`RAIL_GetRadioState(sRailHandle) == RAIL_RF_STATE_IDLE` will return false for any role after thread has been started, (i.e. it only returns true before you type `ifconfig up` in the CLI). This because the radio will subsequently be operating in the other states:

- `RAIL_RF_STATE_RX`
- `RAIL_RF_STATE_TX`
- `RAIL_RF_STATE_ACTIVE`

Unfortunately, this is a limitation in the platform.

From the Silicon Labs RAIL documentation for `RAIL_StartAverageRssi()`

> Starts a non-blocking hardware-based RSSI averaging mechanism. Only a single instance of RSSI averaging can be run at any time and **the radio must be idle to start**.

Reply:
It could be possible to first place the radio in `IDLE` state using `RAIL_Idle()`, just like `otPlatRadioSleep()` does and return it to the previous state after energy scan has completed.

Reply:
Please ignore my deleted comments. The use of `RAIL_GetRadioState()` is incorrect here, it's actually a bitfield. Will update shortly.